File: actin-common/src/main/java/com/hartwig/actin/clinical/feed/questionnaire/Questionnaire.java
package com.hartwig.actin.clinical.feed.questionnaire;

import java.util.List;

import com.hartwig.actin.clinical.datamodel.TumorStage;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class Questionnaire {

    @Nullable
    public abstract String tumorLocation();

    @Nullable
    public abstract String tumorType();

    @Nullable
    public abstract TumorStage stage();

    @Nullable
    public abstract List<String> treatmentHistoryCurrentTumor();

    @Nullable
    public abstract List<String> otherOncologicalHistory();

    @Nullable
    public abstract List<String> nonOncologicalHistory();

    @Nullable
    public abstract Boolean hasMeasurableLesionRecist();

    @Nullable
    public abstract Boolean hasBrainLesions();

    @Nullable
    public abstract Boolean hasActiveBrainLesions();

    @Nullable
    public abstract Boolean hasSymptomaticBrainLesions();

    @Nullable
    public abstract Boolean hasCnsLesions();

    @Nullable
    public abstract Boolean hasActiveCnsLesions();

    @Nullable
    public abstract Boolean hasSymptomaticCnsLesions();

    @Nullable
    public abstract Boolean hasBoneLesions();

    @Nullable
    public abstract Boolean hasLiverLesions();

    @Nullable
    public abstract Integer whoStatus();

    @Nullable
    public abstract Boolean hasSignificantCurrentInfection();

    @Nullable
    public abstract Boolean hasSignificantAberrationLatestECG();

    @Nullable
    public abstract String significantAberrationLatestECG();

    @Nullable
    public abstract List<String> cancerRelatedComplications();

}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/questionnaire/Questionnaire.kt
package com.hartwig.actin.clinical.feed.emc.questionnaire

import com.hartwig.actin.datamodel.clinical.Ecg
import com.hartwig.actin.datamodel.clinical.InfectionStatus
import com.hartwig.actin.datamodel.clinical.TumorStage
import java.time.LocalDate

data class Questionnaire(
    val date: LocalDate,
    val tumorLocation: String?,
    val tumorType: String?,
    val biopsyLocation: String?,
    val stage: TumorStage?,
    val treatmentHistoryCurrentTumor: List<String>?,
    val otherOncologicalHistory: List<String>?,
    val secondaryPrimaries: List<String>?,
    val nonOncologicalHistory: List<String>?,
    val hasMeasurableDisease: Boolean?,
    val hasBrainLesions: Boolean?,
    val hasActiveBrainLesions: Boolean?,
    val hasCnsLesions: Boolean?,
    val hasActiveCnsLesions: Boolean?,
    val hasBoneLesions: Boolean?,
    val hasLiverLesions: Boolean?,
    val otherLesions: List<String>?,
    val ihcTestResults: List<String>?,
    val pdl1TestResults: List<String>?,
    val whoStatus: Int?,
    val unresolvedToxicities: List<String>?,
    val infectionStatus: InfectionStatus?,
    val ecg: Ecg?,
    val complications: List<String>?
) {
    fun isEmpty(): Boolean {
        val allStringsEmpty = listOf(
            tumorLocation,
            tumorType,
            biopsyLocation
        ).all { it.isNullOrEmpty() }

        val allListEmpty = listOf(
            treatmentHistoryCurrentTumor,
            otherOncologicalHistory,
            secondaryPrimaries,
            nonOncologicalHistory,
            otherLesions,
            ihcTestResults,
            pdl1TestResults,
            unresolvedToxicities,
            complications
        ).all { it.isNullOrEmpty() }

        val allBooleanAndObjectFieldsNull = listOf(
            stage,
            hasMeasurableDisease,
            hasBrainLesions,
            hasActiveBrainLesions,
            hasCnsLesions,
            hasActiveCnsLesions,
            hasBoneLesions,
            hasLiverLesions,
            whoStatus,
            infectionStatus,
            ecg
        ).all { it == null }
        return allStringsEmpty && allListEmpty && allBooleanAndObjectFieldsNull
    }
}